### PR TITLE
Ringbuffer - Add support to iovec and add 'reserve'/'commit' for thirdparty writers. (V2)

### PR DESCRIPTION
**Pull request by mbelluzzo**
\+ some methods improvements.

### CODE REVIEW

File: libraries/AP_HAL/utility/RingBuffer.cpp [hunk 1]
```diff
@@ -121,20 +106,81 @@ uint32_t ByteBuffer::peekbytes(uint8_t *data, uint32_t len)
...
         n = len;
     }
 
-    // perform first memcpy
-    memcpy(data, b, n);
-    data += n;
+    iovec[0].data = const_cast<uint8_t *>(b);
+    iovec[0].len = n;
+
+    if (len > n) {
```
Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: nitpick, but I'd say to invert the condition so it's clear you are returning 0, 1 or  2. So here you'd do:
```c++
if (len <= n) {
    return 1;
}
```

and then do the rest below

**Comment by mbelluzzo** [comment]
@lucasdemarchi, pr updated.

File: libraries/AP_HAL/utility/RingBuffer.cpp (hunk 1 above)
Comment (mbelluzzo) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: AAahhh I did in reserve hehehe It had a similar code hehehe fixing it again!

**Comment by mbelluzzo** [comment]
@lucasdemarchi https://github.com/mbelluzzo/ardupilot/tree/stress

**Comment by tridge** [comment]
@mbelluzzo,
thanks so much for writing a stress tester!
I've got this soak testing on a pixracer now

File: libraries/AP_HAL/utility/RingBuffer.cpp [hunk 2]
```diff
@@ -25,7 +25,6 @@ void ByteBuffer::set_size(uint32_t _size)
 {
     uint8_t *oldbuf;
 
-    head = tail = 0;
```
Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: this needs to stay. When you set the size you need to reset head and tail.

Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: what you need to do is to remove them from the constructor.

**Comment by lucasdemarchi** [comment]
@tridge we also saw that this class is actually not used on PX4, only on sitl and qurt. I plan to be a heavy user of it for linux boards, hence why we are improving it.

**Comment by tridge** [comment]
@lucasdemarchi, ahh, sorry! I thought I'd converted from the old macros in AP_HAL_PX4/UARTDriver.cpp to the new RingBuffer but I haven't pushed that yet.
we do use ObjectArray in libraries/GCS_MAVLink, but that isn't built on RingBuffer.
I would like to get rid of all the old macro based ring buffers, but don't let that hold up your work.

File: libraries/AP_HAL/utility/RingBuffer.cpp (hunk 2 above)
Comment (mbelluzzo) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: shit, removed the wrong one :-P

Comment (mbelluzzo) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: fixed

**Comment by mbelluzzo** [comment]
Merged

**Pull request closed without merging**